src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs

```diff
@@ -0,0 +1,140 @@
+        {
+            if (utf8Json == null)
+                throw new ArgumentNullException(nameof(utf8Json));
+
+            if (returnType == null)
+                throw new ArgumentNullException(nameof(returnType));
+
+            return ReadAsync<object>(utf8Json, returnType, options, cancellationToken);
+        }
+
+        private static async ValueTask<TValue> ReadAsync<TValue>(Stream utf8Json, Type returnType, JsonSerializerOptions options = null, CancellationToken cancellationToken = default)
+        {
+            if (options == null)
+                options = s_defaultSettings;
+
+            ReadStack state = default;
+            JsonClassInfo classInfo = options.GetOrAddClass(returnType);
+            state.Current.JsonClassInfo = classInfo;
+            if (classInfo.ClassType != ClassType.Object)
+            {
+                state.Current.JsonPropertyInfo = classInfo.GetPolicyProperty();
+            }
+
+            var readerState = new JsonReaderState(options: options.ReaderOptions);
+
+            int bytesRemaining = 0;
+            int bytesRead;
+
+            // todo: switch to ArrayBuffer implementation to handle and simplify the allocs?
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(options.EffectiveBufferSize);
+            int bufferSize = buffer.Length;
+            int deserializeBufferSize;
+            bool isFinalBlock;
+
+            try
+            {
+                do
+                {
+                    int bytesToRead = bufferSize - bytesRemaining;
+                    bytesRead = await utf8Json.ReadAsync(buffer, bytesRemaining, bytesToRead, cancellationToken).ConfigureAwait(false);
+
+                    deserializeBufferSize = bytesRemaining + bytesRead;
+                    isFinalBlock = (bytesRead == 0);
+
+                    ReadCore(
+                        ref readerState,
+                        isFinalBlock,
+                        buffer,
+                        deserializeBufferSize,
+                        options,
+                        ref state);
+
+                    int bytesConsumed = (int)readerState.BytesConsumed;
+                    bytesRemaining = deserializeBufferSize - bytesConsumed;
+
+                    if (isFinalBlock)
+                    {
+                        break;
+                    }
+
+                    // Check if we need to shift or expand the buffer because there wasn't enough data to complete deserialization.
+                    if (bytesConsumed <= (bufferSize / 2))
+                    {
+                        // We have less than half the buffer available, double the buffer size.
+                        bufferSize = (bufferSize < HalfMaxValue) ? bufferSize * 2 : int.MaxValue;
+
+                        byte[] dest = ArrayPool<byte>.Shared.Rent(bufferSize);
```

Comment:
Renting `int.MaxValue` would throw OOM. Should we cap it to something smaller (either 2GB, or maybe `int.MaxValue - 56`)?

Reply:
Do we have a policy on this elsewhere? It seems like a common scenario.